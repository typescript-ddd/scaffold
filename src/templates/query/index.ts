export * from "./query.generator";
export * from "./query.template";
export * from "./query.types";
