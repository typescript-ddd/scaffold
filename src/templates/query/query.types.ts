import {
  TemplatePropertiesOption,
  TemplatePropertyValue,
  TemplateStringOption,
} from "../shared/template";

export type QueryTemplateValues = {
  entityName: string;
  actionName: string;
  properties: TemplatePropertyValue[];
};

export type QueryTemplateOptions = {
  entityName: TemplateStringOption;
  actionName: TemplateStringOption;
  properties: TemplatePropertiesOption;
};
