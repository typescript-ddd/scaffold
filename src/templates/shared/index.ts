export * from "./context-builder";
export * from "./generate-context";
export * from "./template";