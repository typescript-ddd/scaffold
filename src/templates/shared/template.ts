import { GenerateContext } from "./generate-context";

export type TemplateEnum = Record<string, string>;
export type TemplatePropertyValueType =
  | "string"
  | "number"
  | "boolean"
  | "date";

export type TemplateEnumOption = {
  name: string;
  label: string;
  description: string;
  enum: TemplateEnum;
  optionType: "enum";
};

export type TemplateStringOption = {
  name: string;
  label: string;
  description: string;
  optionType: "string";
};

export type TemplateTokenStringOption = {
  name: string;
  label: string;
  description: string;
  optionType: "tokenString";
};

export type TemplateBooleanOption = {
  name: string;
  label: string;
  description: string;
  optionType: "boolean";
};

export type TemplateNumberOption = {
  name: string;
  label: string;
  description: string;
  optionType: "number";
};

export type TemplateDateOption = {
  name: string;
  label: string;
  description: string;
  optionType: "date";
};

export type TemplatePropertyOption = {
  name: string;
  valueType: TemplatePropertyValueType;
  nullable?: boolean;
  prop?: string;
};

export type TemplatePropertyValue = {
  name: string;
  valueType: string;
  nullable?: boolean;
  prop?: string;
};

export type TemplatePropertiesOption = {
  name: string;
  label: string;
  description: string;
  optionType: "properties";
};

export type TemplateOption =
  | TemplateEnumOption
  | TemplateStringOption
  | TemplateBooleanOption
  | TemplateNumberOption
  | TemplateDateOption
  | TemplatePropertiesOption
  | TemplateTokenStringOption;

export type Maybe<T> = T | null | undefined;

export interface Template<
  TValues extends { [K in keyof TOptions]: any},
  TOptions extends { [key: string]: TemplateOption }
> {
  name: string;
  description: string;
  options: TOptions;
  defaultValues: TValues;
  generate(values: TValues, context: GenerateContext): string;
}