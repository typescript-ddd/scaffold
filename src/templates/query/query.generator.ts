import { Project, Scope } from "ts-morph";
import { strings } from "../../utils";
import { Chunk, GenerateContext } from "../shared";
import { QueryTemplateValues } from "./query.types";

export const generateQuery = (
  values: QueryTemplateValues,
  context: GenerateContext,
  chunkName?: string
): Chunk => {
  const { entityName, actionName = "Find", properties } = values;
  const queryName = `${strings.capitalize(actionName)}${strings.capitalize(
    entityName
  )}Query`;
  const {
    fileName = `${strings.kebab("find", entityName, "query")}.ts`,
    projectPath = [],
  } = context.currentFile || {};
  const project = new Project();
  const sourceFile = project.createSourceFile(
    fileName,
    `
export class ${queryName} implements Query {}    
`
  );

  context.addImportDeclaration({
    moduleSpecifier: `@typescript-ddd/core`,
    namedImports: ["Query"],
  });

  context.writeSourceFileImports(sourceFile);

  const classDeclaration = sourceFile.getClass(queryName);
  classDeclaration.addJsDoc({
    description: `Represents ${strings.describe(queryName)}.`,
  });

  const constructorDeclaration = classDeclaration.addConstructor({
    parameters: [
      ...properties.map((property) => ({
        name: property.name,
        type: property.valueType,
        hasQuestionToken: property.nullable,
        scope: Scope.Public,
        isReadonly: true,
      })),
    ],
  });
  constructorDeclaration.setBodyText(`// TODO: Add validation`);
  constructorDeclaration.addJsDoc({
    description: `Initializes a new instance of ${strings.describe(
      queryName
    )}.`,
    tags: properties.map((property) => ({
      tagName: "param",
      text: `{${property.valueType}} ${property.name} - The ${property.name}.`,
    })),
  });

  return {
    name: chunkName || "Query",
    fileName,
    projectPath,
    content: sourceFile.getFullText(),
    values,
  };
};
