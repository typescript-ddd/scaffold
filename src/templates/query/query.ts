import { Project, Scope } from "ts-morph";
import { strings } from "../../utils";
import { GenerateContext } from "../shared";

export type GenerateQueryOptions = {
  entityName: string;
  actionName?: string;
  properties: {
    name: string;
    type: string;
    nullable?: boolean;
    description?: string;
  }[];
};

export const generateQuery = (
  options: GenerateQueryOptions,
  context: GenerateContext
) => {
  const { entityName, actionName = "Find", properties } = options;
  const queryName = `${strings.capitalize(actionName)}${strings.capitalize(
    entityName
  )}Query`;

  const project = new Project();
  const sourceFile = project.createSourceFile(
    `${context.fileName}.ts`,
    `
export class ${queryName} implements Query {}    
`
  );

  context.addImportDeclaration({
    moduleSpecifier: `@typescript-ddd/core`,
    namedImports: ["Query"],
  });

  context.writeSourceFileImports(sourceFile);
  
  const classDeclaration = sourceFile.getClass(queryName);
  classDeclaration.addJsDoc({
    description: `Represents ${strings.describe(queryName)}.`,
  });

  const constructorDeclaration = classDeclaration.addConstructor({
    parameters: [
      ...properties.map((property) => ({
        name: property.name,
        type: property.type,
        hasQuestionToken: property.nullable,
        scope: Scope.Public,
        isReadonly: true,
      })),
    ],
  });
  constructorDeclaration.setBodyText(`// TODO: Add validation`);
  constructorDeclaration.addJsDoc({
    description: `Initializes a new instance of ${strings.describe(
      queryName
    )}.`,
    tags: properties.map((property) => ({
      tagName: "param",
      text: `{${property.type}} ${property.name} ${
        property.description ? `- ${property.description}` : ""
      }`,
    })),
  });

  let output = sourceFile.getFullText();

  return output;
};
