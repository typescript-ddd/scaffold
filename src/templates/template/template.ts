import { Project, Scope } from "ts-morph";
import { strings } from "../../utils";
import { GenerateContext } from "../shared";

export type GenerateTemplateOptions = {
  name: string;
};

export const generateTemplate = (
  options: GenerateTemplateOptions,
  context: GenerateContext
) => {
  const { name } = options;
  const className = strings.capitalize(name);

  const project = new Project();
  const sourceFile = project.createSourceFile(
    `${context.fileName}.ts`,
    `
export class ${className} {}    
`
  );

  context.addImportDeclaration({
    moduleSpecifier: `@typescript-ddd/core`,
    namedImports: ["Entity"],
  });

  context.writeSourceFileImports(sourceFile);

  const classDeclaration = sourceFile.getClass(className);
  classDeclaration.addJsDoc({
    description: `${strings.describe(name)}.`,
  });

  const constructorDeclaration = classDeclaration.addConstructor({
    scope: Scope.Private,
    parameters: [{ name: "value", type: "string" }],
  });
  constructorDeclaration.setBodyText(`super(value);`);
  constructorDeclaration.addJsDoc({
    description: `Initializes a new instance of ${strings.describe(
      name
    )} identifier.`,
    tags: [{ tagName: "param", text: "value The value of the identifier." }],
  });

  let output = sourceFile.getFullText();

  return output;
};
