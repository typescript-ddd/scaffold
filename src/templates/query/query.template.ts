import { GenerateContext } from "../shared";
import { Template } from "../shared/template";
import { generateQuery } from "./query.generator";
import {
  QueryTemplateOptions,
  QueryTemplateValues,
} from "./query.types";

export class QueryTemplate
  implements
    Template<QueryTemplateValues, QueryTemplateOptions>
{
  readonly name = "Query";
  readonly description = "Generates a query";
  readonly options: QueryTemplateOptions = {
    entityName: {
      name: "entityName",
      label: "Entity Name",
      description: "The name of the entity",
      optionType: "string",
    },
    actionName: {
        name: "actionName",
        label: "Action Name",
        description: "The name of the action",
        optionType: "string",
    },
    properties: {
        name: "properties",
        label: "Properties",
        description: "The properties of the entity",
        optionType: "properties",
    }
  };
  readonly defaultValues = {
    entityName: "User",
    actionName: "find",
    properties: [],
  };
  generate(values: QueryTemplateValues, context: GenerateContext): string {
    return generateQuery(values, context);
  }
}
