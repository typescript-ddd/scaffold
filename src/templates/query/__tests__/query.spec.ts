import { ContextBuilder, GenerateContext } from "../../shared";
import { QueryTemplate } from "../query.template";

describe("query", () => {
  let context: GenerateContext;
  let template: QueryTemplate;

  beforeEach(() => {
    context = ContextBuilder.build("@src");
    template = new QueryTemplate();
  });

  it("Should render the query template", () => {
    context.addImportDeclaration({
      moduleSpecifier: context.resolveDir("domain"),
      namedImports: ["UserId"],
    });

    const output = template.generate(
      {
        entityName: "User",
        actionName: "find",
        properties: [
          { name: "id", valueType: "UserId" },
        ],
      },
      context
    );

    expect(output).toMatchSnapshot();
  });
});
