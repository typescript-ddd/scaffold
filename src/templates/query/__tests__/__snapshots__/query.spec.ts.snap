// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`query Should render the query template 1`] = `
"import { Query } from "@typescript-ddd/core";
import { UserId } from "@src/domain";

/** Represents a find user query. */
export class FindUserQuery implements Query {
    /**
     * Initializes a new instance of a find user query.
     * @param {UserId} id - The id.
     */
    constructor(public readonly id: UserId) {
        // TODO: Add validation
    }
}    
"
`;
